lib/shared/sdk/usbboot/protocol.js

```diff
@@ -0,0 +1,314 @@
+ */
+const USB_REQUEST_CODE_GET_STATUS = 0
+
+/**
+ * @summary The maximum buffer length of a usbboot message
+ * @type {Number}
+ * @constant
+ */
+const USBBOOT_MESSAGE_MAX_BUFFER_LENGTH = 0xffff
+
+/**
+ * @summary The delay to wait between each USB read/write operation
+ * @type {Number}
+ * @constant
+ * @description
+ * The USB bus seems to hang if we execute many operations at
+ * the same time.
+ */
+const USB_REQUEST_DELAY_MS = 1000
+
+/**
+ * @summary The timeout for USB bulk transfers, in milliseconds
+ * @type {Number}
+ * @constant
+ */
+const USB_BULK_TRANSFER_TIMEOUT_MS = 1000
+
+/**
+ * @summary The size of the usbboot file message
+ * @type {Number}
+ * @constant
+ */
+exports.FILE_MESSAGE_LENGTH = FILE_MESSAGE_COMMAND_LENGTH + FILE_MESSAGE_FILE_NAME_LENGTH
+
+/**
+ * @summary The "get file size" file message command name
+ * @type {String}
+ * @constant
+ */
+exports.FILE_MESSAGE_COMMAND_GET_FILE_SIZE = 'GetFileSize'
+
+/**
+ * @summary The "read file" file message command name
+ * @type {String}
+ * @constant
+ */
+exports.FILE_MESSAGE_COMMAND_READ_FILE = 'ReadFile'
+
+/**
+ * @summary The "done" file message command name
+ * @type {String}
+ * @constant
+ */
+exports.FILE_MESSAGE_COMMAND_DONE = 'Done'
+
+/**
+ * @summary The usbboot return code that represents success
+ * @type {Number}
+ * @constant
+ */
+exports.RETURN_CODE_SUCCESS = 0
+
+/**
+ * @summary The buffer length of the return code message
+ * @type {Number}
+ * @constant
+ */
+exports.RETURN_CODE_LENGTH = 4
+
+exports.sendBufferSize = (device, size) => {
+  return usb.performControlTransfer(device, {
+    bmRequestType: usb.LIBUSB_REQUEST_TYPE_VENDOR,
+    bRequest: USB_REQUEST_CODE_GET_STATUS,
+    data: NULL_BUFFER,
+
+    /* eslint-disable no-bitwise */
+    wValue: size & USBBOOT_MESSAGE_MAX_BUFFER_LENGTH,
+    wIndex: size >> 16
+    /* eslint-enable no-bitwise */
+  })
+}
+
+/**
+ * @summary Write a buffer to an OUT endpoint
+ * @function
+ * @private
+ *
+ * @param {Object} device - device
+ * @param {Object} endpoint - endpoint
+ * @param {Buffer} buffer - buffer
+ * @returns {Promise}
+ *
+ * @example
+ * usbboot.scan().then((devices) => {
+ *   return protocol.write(devices[0], devices[0].interface(0).endpoint(1), Buffer.alloc(1)).then(() => {
+ *     console.log('Done!')
+ *   })
+ * })
+ */
+exports.write = (device, endpoint, buffer) => {
+  return exports.sendBufferSize(device, buffer.length)
+
+    // We get LIBUSB_TRANSFER_STALL sometimes
+    // in future bulk transfers without this
+    .delay(USB_REQUEST_DELAY_MS)
+
+    .then(() => {
+      return Bluebird.fromCallback((callback) => {
+        endpoint.timeout = USB_BULK_TRANSFER_TIMEOUT_MS
+        endpoint.transfer(buffer, callback)
+      })
+    })
+}
+
+exports.sendErrorSignal = (device) => {
+  // The original implementation sends a control transfer
+  // with the size of a null buffer, and then the null
+  // buffer itself, however sending the actual null buffer
```

Comment:
> The original implementation sends ... and then the null buffer itself

Are you sure? Doesn't https://github.com/raspberrypi/usbboot/blob/master/main.c#L149 mean that the NULL buffer _doesn't_ get written?